### PR TITLE
Add Email OTP as a service

### PR DESCRIPTION
**Pull request by maneeshaindrachapa**
Git Issue: https://github.com/wso2/product-is/issues/12484

### Purpose
To expose the Email OTP generation and validation capability as a fully decoupled OSGI service.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -2,4 +2,49 @@ Welcome to the WSO2 Identity Server (IS) EmailOTP authenticator.
...
+    ....
+    Please use this One-Time Password {{OTPCode}}
+    ....
+    </configuration>
+```
+
+5.Restart the server.
+
+## REST API support
+REst API support for this OSGI service can be found in the below git repository.
```
Comment (nipunthathsara) [inline README.md]: REST

File: component/common/pom.xml [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+        <dependency>
+            <groupId>org.testng</groupId>
+            <artifactId>testng</artifactId>
+            <version>${testng.version}</version>
+            <scope>test</scope>
+        </dependency>
+        <dependency>
+            <groupId>org.powermock</groupId>
+            <artifactId>powermock-module-junit4</artifactId>
+            <version>${powermock.version}</version>
```
Comment (nipunthathsara) [inline component/common/pom.xml]: MOve versions to the parent pom. as we have done the same in the SMS common.

File: component/common/pom.xml [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+            <artifactId>powermock-api-mockito</artifactId>
+            <version>${powermock.version}</version>
+            <scope>test</scope>
+        </dependency>
+        <!--        <dependency>-->
+        <!--            <groupId>org.mockito</groupId>-->
+        <!--            <artifactId>mockito-all</artifactId>-->
+        <!--            <version>${mockito.version}</version>-->
+        <!--            <scope>test</scope>-->
+        <!--        </dependency>-->
```
Comment (nipunthathsara) [inline component/common/pom.xml]: Remove if we don't want this.

File: component/common/pom.xml [hunk 4]
```diff
@@ -0,0 +1,125 @@
...
+                            org.apache.commons.logging.*; version="1.0.4",
+                            org.osgi.framework,
+                            org.osgi.service.component; version="${osgi.service.import.version.range}",
+                            *;resolution:=optional
+                        </Import-Package>
+                        <Export-Package>
+                            !org.wso2.carbon.extension.identity.emailotp.common.internal,
+                            org.wso2.carbon.extension.identity.emailotp.common.*; version="${project.version}",
+                        </Export-Package>
+                        <DynamicImport-Package>*</DynamicImport-Package>
```
Comment (nipunthathsara) [inline component/common/pom.xml]: Remove dynamic import and add all imports.

File: component/common/pom.xml [hunk 5]
```diff
@@ -0,0 +1,125 @@
...
+                    </instructions>
+                </configuration>
+            </plugin>
+        </plugins>
+    </build>
+</project>
```
Comment (nipunthathsara) [inline component/common/pom.xml]: New line.

File: component/common/src/main/java/org/wso2/carbon/extension/identity/emailotp/common/EmailOtpService.java [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+            EmailOtpException;
+
+    /**
+     * This method will generate an OTP and send an EMAIL notification.
+     *
+     * @param userId
+     * @return {@link GenerationResponseDTO}
+     * @throws EmailOtpException
+     */
+    GenerationResponseDTO generateEmailOTP(String userId) throws EmailOtpException;
```
Comment (nipunthathsara) [inline component/common/src/main/java/org/wso2/carbon/extension/identity/emailotp/common/EmailOtpService.java]: Describe all the fields including exceptions.

File: component/common/src/main/java/org/wso2/carbon/extension/identity/emailotp/common/dto/GenerationResponseDTO.java [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.wso2.carbon.extension.identity.emailotp.common.dto;
+
+/**
+ * This class represents a model of to generate EMAIL OTP request.
```
Comment (nipunthathsara) [inline component/common/src/main/java/org/wso2/carbon/extension/identity/emailotp/common/dto/GenerationResponseDTO.java]: Not request, response.

File: component/common/src/main/java/org/wso2/carbon/extension/identity/emailotp/common/dto/SessionDTO.java [hunk 8]
```diff
@@ -0,0 +1,134 @@
...
+
+package org.wso2.carbon.extension.identity.emailotp.common.dto;
+
+import java.io.Serializable;
+import java.util.Objects;
+
+/**
+ * Object used to save the data in the database
+ */
+
```
Comment (nipunthathsara) [inline component/common/src/main/java/org/wso2/carbon/extension/identity/emailotp/common/dto/SessionDTO.java]: Addintional new line.

File: component/common/src/main/java/org/wso2/carbon/extension/identity/emailotp/common/dto/ValidationResponseDTO.java [hunk 9]
```diff
@@ -0,0 +1,72 @@
...
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.wso2.carbon.extension.identity.emailotp.common.dto;
+
+/**
+ * This class represents a model of to validate EMAIL OTP request.
```
Comment (nipunthathsara) [inline component/common/src/main/java/org/wso2/carbon/extension/identity/emailotp/common/dto/ValidationResponseDTO.java]: Response.

File: component/common/src/main/java/org/wso2/carbon/extension/identity/emailotp/common/exception/EmailOtpClientException.java [hunk 10]
```diff
@@ -0,0 +1,44 @@
...
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.wso2.carbon.extension.identity.emailotp.common.exception;
+
+/**
+ * EMAIL OTP client exception
```
Comment (nipunthathsara) [inline component/common/src/main/java/org/wso2/carbon/extension/identity/emailotp/common/exception/EmailOtpClientException.java]: Period.

Comment (nipunthathsara) [reply component/common/src/main/java/org/wso2/carbon/extension/identity/emailotp/common/exception/EmailOtpClientException.java]: And the word 'Email' is not an acronym. Therefore, better to use as Email in all comments and logs.

File: component/common/src/main/java/org/wso2/carbon/extension/identity/emailotp/common/exception/EmailOtpServerException.java [hunk 11]
```diff
@@ -0,0 +1,43 @@
...
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.wso2.carbon.extension.identity.emailotp.common.exception;
+
+/**
+ * EMAIL OTP server exception
```
Comment (nipunthathsara) [inline component/common/src/main/java/org/wso2/carbon/extension/identity/emailotp/common/exception/EmailOtpServerException.java]: Period. Please check and fix all other places.

File: component/common/src/main/java/org/wso2/carbon/extension/identity/emailotp/common/util/OneTimePasswordUtils.java [hunk 12]
```diff
@@ -0,0 +1,247 @@
...
+    /**
+     * This method uses the JCE to provide the HMAC-SHA-1
+     * algorithm. HMAC computes a Hashed Message Authentication Code and in this
+     * case SHA1 is the hash algorithm used.
+     *
+     * @param keyBytes Bytes to use for the HMAC-SHA-1 key.
+     * @param text     Message or text to be authenticated.
+     * @throws NoSuchAlgorithmException If no provider makes either HmacSHA1 or HMAC-SHA-1 digest
+     *                                  algorithms available.
+     * @throws InvalidKeyException      The secret provided was not a valid HMAC-SHA-1 key.
```
Comment (nipunthathsara) [inline component/common/src/main/java/org/wso2/carbon/extension/identity/emailotp/common/util/OneTimePasswordUtils.java]: Indent and Align as in other methods.

File: component/common/src/main/java/org/wso2/carbon/extension/identity/emailotp/common/util/Utils.java [hunk 13]
```diff
@@ -0,0 +1,181 @@
...
+ * Util functions for EMAIL OTP service.
+ */
+public class Utils {
+
+    private static final Log log = LogFactory.getLog(Utils.class);
+
+    /**
+     * Read configurations and populate {@link ConfigsDTO} object.
+     *
+     * @throws EmailOtpServerException Throws upon an issue on while reading configs.
```
Comment (nipunthathsara) [inline component/common/src/main/java/org/wso2/carbon/extension/identity/emailotp/common/util/Utils.java]: "on while"

**Review by nipunthathsara — CHANGES_REQUESTED** [review]

File: README.md (hunk 1 above)
Comment (maneeshaindrachapa) [reply README.md]: Fixed in 5fe267f8e9157aeb87e20a328965679356db8a84

File: component/common/pom.xml (hunk 2 above)
Comment (maneeshaindrachapa) [reply component/common/pom.xml]: Fixed in 5fe267f8e9157aeb87e20a328965679356db8a84

File: component/common/pom.xml (hunk 3 above)
Comment (maneeshaindrachapa) [reply component/common/pom.xml]: Fixed in 5fe267f8e9157aeb87e20a328965679356db8a84

File: component/common/pom.xml (hunk 4 above)
Comment (maneeshaindrachapa) [reply component/common/pom.xml]: Fixed in 5fe267f8e9157aeb87e20a328965679356db8a84

File: component/common/pom.xml (hunk 5 above)
Comment (maneeshaindrachapa) [reply component/common/pom.xml]: Fixed in 5fe267f8e9157aeb87e20a328965679356db8a84

File: component/common/src/main/java/org/wso2/carbon/extension/identity/emailotp/common/dto/GenerationResponseDTO.java (hunk 7 above)
Comment (maneeshaindrachapa) [reply component/common/src/main/java/org/wso2/carbon/extension/identity/emailotp/common/dto/GenerationResponseDTO.java]: Fixed in 5fe267f8e9157aeb87e20a328965679356db8a84

File: component/common/src/main/java/org/wso2/carbon/extension/identity/emailotp/common/dto/ValidationResponseDTO.java (hunk 9 above)
Comment (maneeshaindrachapa) [reply component/common/src/main/java/org/wso2/carbon/extension/identity/emailotp/common/dto/ValidationResponseDTO.java]: Fixed in 5fe267f8e9157aeb87e20a328965679356db8a84

File: component/common/src/main/java/org/wso2/carbon/extension/identity/emailotp/common/exception/EmailOtpClientException.java (hunk 10 above)
Comment (maneeshaindrachapa) [reply component/common/src/main/java/org/wso2/carbon/extension/identity/emailotp/common/exception/EmailOtpClientException.java]: Fixed in 5fe267f8e9157aeb87e20a328965679356db8a84

File: component/common/src/main/java/org/wso2/carbon/extension/identity/emailotp/common/exception/EmailOtpServerException.java (hunk 11 above)
Comment (maneeshaindrachapa) [reply component/common/src/main/java/org/wso2/carbon/extension/identity/emailotp/common/exception/EmailOtpServerException.java]: Fixed in 5fe267f8e9157aeb87e20a328965679356db8a84

File: component/common/src/main/java/org/wso2/carbon/extension/identity/emailotp/common/util/OneTimePasswordUtils.java (hunk 12 above)
Comment (maneeshaindrachapa) [reply component/common/src/main/java/org/wso2/carbon/extension/identity/emailotp/common/util/OneTimePasswordUtils.java]: Fixed in 5fe267f8e9157aeb87e20a328965679356db8a84

File: component/common/src/main/java/org/wso2/carbon/extension/identity/emailotp/common/EmailOtpService.java (hunk 6 above)
Comment (maneeshaindrachapa) [reply component/common/src/main/java/org/wso2/carbon/extension/identity/emailotp/common/EmailOtpService.java]: Fixed in af803a284224d961eb0a744e751122f917b9b0af

File: component/common/src/main/java/org/wso2/carbon/extension/identity/emailotp/common/dto/SessionDTO.java (hunk 8 above)
Comment (maneeshaindrachapa) [reply component/common/src/main/java/org/wso2/carbon/extension/identity/emailotp/common/dto/SessionDTO.java]: Fixed in af803a284224d961eb0a744e751122f917b9b0af

File: component/common/src/main/java/org/wso2/carbon/extension/identity/emailotp/common/util/Utils.java (hunk 13 above)
Comment (maneeshaindrachapa) [reply component/common/src/main/java/org/wso2/carbon/extension/identity/emailotp/common/util/Utils.java]: Fixed in af803a284224d961eb0a744e751122f917b9b0af

**Pull request merged**